File: Recusrion/subsequenses_with_k_sum.py
"""
this function for retrung any subsequence with a specific sum
"""
def rec_any(idx , ds: list[int] , arr: list[int], sum , tot, n):
    
    if idx>=n:
        if sum == tot:
            print (*ds)
            return True
        
        return False
    
    ds.append(arr[idx])
    sum+=arr[idx]
    if  rec_any(idx+1, ds, arr, sum, tot, n) == True:
        return True
    
    ds.pop()
    sum-=arr[idx]
    if  rec_any(idx+1, ds, arr, sum, tot, n) == True:
        return True
    
    
    return False
    
    
"""
this function for printing all subsequences with a specific sum
"""
def rec(idx , ds: list[int] , arr: list[int], sum , tot, n):
    
    if idx>=n:
        if sum == tot:
            print (*ds)
        return 
    
    ds.append(arr[idx])
    sum+=arr[idx]
    rec(idx+1, ds, arr, sum, tot, n)
    ds.pop()
    sum-=arr[idx]
    rec(idx+1, ds, arr, sum, tot, n)

            

if __name__ == "__main__":
    
    a = [1,2,1]
    tot=2
    n = 3
    ds = []
    rec_any(0,ds,a,0,tot,n)

File: Recusrion/subsequenses_with_k_sum.cpp
#include<bits/stdc++.h>
using namespace std;
/*
this function for retrung any subsequence with a specific sum
*/
bool rec_any(int idx , vector<int>&ds ,int sum , int tot , int n , int arr[])
{
    if(idx >= n)
    {
        if(sum == tot)
        {
            for (auto it : ds)
            {
                cout<<it<<" ";
            }
            cout<<endl;
            return true;
        }
        return false;
    }
    ds.push_back(arr[idx]); //take
    sum+=arr[idx];
    if(rec_any(idx+1,ds,sum,tot,n,arr) == true){
        return true;
    }

    ds.pop_back();  //not take
    sum-=arr[idx];
    if(rec_any(idx+1,ds,sum,tot,n,arr) == true)
    {
        return true;
    }

    return false;


}


/*
this function for printing all subsequences with a specific sum
*/

void rec(int idx , vector<int>&ds ,int sum , int tot , int n , int arr[])
{
    if(idx >= n)
    {
        if(sum == tot)
        {
            for (auto it : ds)
            {
                cout<<it<<" ";
            }
            cout<<endl;
        }
        return;
    }
    ds.push_back(arr[idx]); //take
    sum+=arr[idx];
    rec(idx+1,ds,sum,tot,n,arr);
    ds.pop_back();  //not take
    sum-=arr[idx];
    rec(idx+1,ds,sum,tot,n,arr);


}
int main()
{  
    int n = 4;
    int a[] = {1,2,1,0};
    int tot = 2;
    vector<int>ds;
    rec_any(0,ds,0,tot,n,a);

}